File: CarePackage/Software/NodeJsLts.cs
namespace CarePackage.Software;

public class NodeJsLts : BaseSoftware
{
    public override string           Key            { get; set; } = "nodejs-lts";
    public override string           Name           { get; set; } = "Node.js (LTS)";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Runtime;
    public override string           DownloadName   { get; set; } = "node-lts-x64.msi";
    public override bool             IsArchive      { get; set; } = false;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.nodejs;
    public override string           Homepage       { get; set; } = "https://nodejs.org";
    
    private readonly HttpClient _http;

    public NodeJsLts(IServiceProvider services)
    {
        _http = services.GetKeyedService<HttpClient>("MimicBrowser")!;
    }

    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var res = await _http.GetAsync("https://nodejs.org/en", ct);
        
        res.EnsureSuccessStatusCode();
        
        var versionPattern = new Regex(@"dist\/(v\d{2}\.\d{1,2}\.\d{1,})\/node");
        var html           = await res.Content.ReadAsStringAsync(ct);
        var matches        = versionPattern.Matches(html);
        var version = matches
                      .OrderBy(m => new Version(m.Groups[1].Value[1..]))
                      .FirstOrDefault()
                      ?.Groups[1].Value;
        
        DownloadUrlResolveException.ThrowIf(version is null);

        return $"https://nodejs.org/dist/{version}/node-{version}-x64.msi";
    }
}

File: packages/app/src/services/software/definitions/variants/nodejsLts.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'nodejs-lts';
	public name = 'Node.js (LTS)';
	public category = [SoftwareCategory.Development, SoftwareCategory.Runtime];
	public downloadName = 'node-lts-x64.msi';
	public shouldCacheUrl = true;
	public icon = 'nodejs.png';
	public homepage = 'https://nodejs.org';
	public parent = 'nodejs';

	public async resolveDownloadUrl() {
		const res = await fetch('https://nodejs.org/dist/index.json');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const json        = await res.json() as any[];
		const { version } = json.find(r => r.lts !== false);

		return ok(`https://nodejs.org/dist/${version}/node-${version}-x64.msi`);
	}
}
